File: epi_judge_cpp/string_integer_interconversion.cc
#include <string>

#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
using std::string;

string IntToString(int x)
{
    string res;
    bool is_negative = x < 0;

    // cannot convert x to positive in case x is int min

    do {
        res.push_back('0' + abs(x % 10));
        x /= 10;
    } while (x);

    if (is_negative) {
        res.push_back('-');
    }
    
    return {res.rbegin(), res.rend()};
}
int StringToInt(const string &s)
{
    int res = 0;

    int sign = s[0] == '-' ? -1 : 1;

    for (char c : s) {
        if (c == '+' || c == '-') {
            continue;
        }

        res *= 10;
        res += c - '0';
    }

    return sign * res;
}
void Wrapper(int x, const string &s)
{
    if (stoi(IntToString(x)) != x)
    {
        throw TestFailure("Int to string conversion failed");
    }

    if (StringToInt(s) != x)
    {
        throw TestFailure("String to int conversion failed");
    }
}

int main(int argc, char *argv[])
{
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"x", "s"};
    return GenericTestMain(args, "string_integer_interconversion.cc",
                           "string_integer_interconversion.tsv", &Wrapper,
                           DefaultComparator{}, param_names);
}

File: epi_judge_python/string_integer_interconversion.py
from test_framework import generic_test
from test_framework.test_failure import TestFailure


def int_to_string(x: int) -> str:
    sign = '' if x >= 0 else '-'
    x = abs(x)
    num_str = ''

    while x:
        num_str += chr(x % 10 + 48)
        x //= 10

    num_str = ''.join(reversed(num_str))
    num_str = '0' if num_str == '' else num_str
    return sign + num_str


def string_to_int(s: str) -> int:
    n = len(s)
    sign = 1
    i = 0
    num = 0
    if s[i] == '+':
        i += 1
    elif s[i] == '-':
        sign = -1
        i += 1

    while i < n:
        num *= 10
        num += ord(s[i]) - 48
        i += 1

    return sign * num


def wrapper(x, s):
    if int(int_to_string(x)) != x:
        raise TestFailure('Int to string conversion failed')
    if string_to_int(s) != x:
        raise TestFailure('String to int conversion failed')


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('string_integer_interconversion.py',
                                       'string_integer_interconversion.tsv',
                                       wrapper))
